(feat) integrate smartctl exporter (#5164)
grafana/alloy: docs/sources/reference/components/prometheus/prometheus.exporter.smartctl.md

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.smartctl/
description: Learn about prometheus.exporter.smartctl
labels:
  stage: general-availability
  products:
    - oss
  tags:
    - text: Community
      tooltip: This component is developed, maintained, and supported by the Alloy user community.
title: prometheus.exporter.smartctl
---

# `prometheus.exporter.smartctl`

{{< docs/shared lookup="stability/community.md" source="alloy" version="<ALLOY_VERSION>" >}}

The `prometheus.exporter.smartctl` component collects SMART disk health metrics from storage devices on the local system.

> It uses the `smartctl` from the smartmontools package to query device health, temperature, power-on hours, and other attributes.

Reviewer: ```suggestion
It uses the `smartctl` command from the `smartmontools` package to query device health, temperature, power-on hours, and other attributes.
```

| `smartctl_path`     | `string`       | Path to the smartctl binary.                               | `"/usr/sbin/smartctl"` | no       |
| `scan_interval`     | `duration`     | How often to poll smartctl for device data.                | `"60s"`                | no       |
| `rescan_interval`   | `duration`     | How often to rescan for new or removed devices.            | `"10m"`                | no       |
| `devices`           | `list(string)` | List of specific devices to monitor.                       | `[]`                   | no       |
| `device_exclude`    | `string`       | Regex pattern to exclude devices from automatic scanning.  | `""`                   | no       |
| `device_include`    | `string`       | Regex pattern to include only matching devices.            | `""`                   | no       |
| `scan_device_types` | `list(string)` | Device types to scan (for example, `sat`, `scsi`, `nvme`). | `[]`                   | no       |
| `powermode_check`   | `string`       | Power mode threshold to skip checking devices.             | `"standby"`            | no       |

> The `smartctl_path` must point to a smartctl binary version 7.0 or later with JSON output support.

Reviewer: ```suggestion
The `smartctl_path` must point to a `smartctl` binary version 7.0 or later with JSON output support.
```

The `scan_device_types` argument controls which device types to scan.
Common values include:

- `sat`: SATA devices
- `scsi`: SAS and SCSI devices
- `nvme`: NVMe devices
- `auto`: Auto-detect device type, which is the default when not specified

The `powermode_check` argument determines when to skip checking devices based on their power state to avoid waking sleeping drives.
Valid values are:

- `never`: Always check devices regardless of power state
- `sleep`: Skip devices in sleep mode or deeper
- `standby`: Default, skip devices in standby mode or deeper
- `idle`: Skip devices in idle mode or deeper

## Blocks

The `prometheus.exporter.smartctl` component doesn't support any blocks. You can configure this component with arguments.

## Exported fields

{{< docs/shared lookup="reference/components/exporter-component-exports.md" source="alloy" version="<ALLOY_VERSION>" >}}

## Component health

`prometheus.exporter.smartctl` is only reported as unhealthy when given an invalid configuration.
In those cases, exported fields retain their last healthy values.

The `smartctl` command execution failures don't affect component health.

> Instead, the component sets the `smartctl_device_scrape_success` metric to 0 for devices that fail to scrape.

Reviewer: ```suggestion
Instead, the component sets the `smartctl_device_scrape_success` metric to `0` for devices that fail to scrape.
```

| ------------------------------- | ------- | ----------------------------- |
| `smartctl_device_bytes_read`    | counter | Total bytes read from device  |
| `smartctl_device_bytes_written` | counter | Total bytes written to device |

### Error metrics

| Metric name                           | Type    | Description                 |
| ------------------------------------- | ------- | --------------------------- |
| `smartctl_device_media_errors`        | counter | Device media errors         |
| `smartctl_device_num_err_log_entries` | counter | Number of error log entries |

### NVMe-specific metrics

| Metric name                                 | Type    | Description                          |
| ------------------------------------------- | ------- | ------------------------------------ |
| `smartctl_device_percentage_used`           | counter | Percentage of device lifespan used   |
| `smartctl_device_available_spare`           | counter | Available spare capacity percentage  |
| `smartctl_device_available_spare_threshold` | counter | Available spare threshold percentage |
| `smartctl_device_critical_warning`          | counter | Critical warning status              |

> ### SMART attributes (ATA/SATA only)

Reviewer: ```suggestion
### SMART attributes for ATA and SATA only
```

| Metric name                 | Type  | Description                                                |
| --------------------------- | ----- | ---------------------------------------------------------- |
| `smartctl_device_attribute` | gauge | SMART attribute values: normalized, raw, worst, threshold  |

### Scrape metrics

| Metric name                               | Type  | Description                                     |
| ----------------------------------------- | ----- | ----------------------------------------------- |
| `smartctl_device_scrape_success`          | gauge | Whether the scrape was successful (1 = success) |

> | `smartctl_device_scrape_duration_seconds` | gauge | Duration of the `smartctl` scrape in seconds    |

Reviewer: ```suggestion
| Metric name                               | Type  | Description                                      |
| ----------------------------------------- | ----- | ------------------------------------------------ |
| `smartctl_device_scrape_success`          | gauge | Device scrape success (1 = success, 0 = failure) |
| `smartctl_device_scrape_duration_seconds` | gauge | Duration of the `smartctl` scrape in seconds     |
```

All device metrics include a `device` label with the device path, for example `/dev/sda`.

The `smartctl_device` metric includes comprehensive device information labels:

- `device`: Device path
- `model_name`: Device model name
- `model_family`: Device model family
- `serial_number`: Device serial number
- `firmware_version`: Firmware version
- `interface`: Interface type (for example, SAT, NVMe)
- `protocol`: Protocol (for example, ATA, NVMe)
- `form_factor`: Form factor (for example, 2.5 inches, M.2)
- `ata_version`: ATA version string
- `sata_version`: SATA version string

The `smartctl_device_attribute` metric includes labels:

- `device`: Device path
- `attribute_id`: SMART attribute ID (for example, 5 for Reallocated Sectors)
- `attribute_name`: SMART attribute name
- `attribute_value_type`: Type of value (normalized, raw, worst, threshold)
- `attribute_flags_short`: Short flags string
- `attribute_flags_long`: Long flags string (comma-separated)

## Prerequisites

Before using `prometheus.exporter.smartctl`, ensure the following requirements are met:

> 1. **smartmontools** version 7.0 or later must be installed on the system:

Reviewer: ```suggestion
1. Install `smartmontools` version 7.0 or later:
```

   - Debian/Ubuntu: `sudo apt-get install smartmontools`
   - RHEL/CentOS: `sudo yum install smartmontools`

>    - macOS: `brew install smartmontools`

Author: I'm not really familiar with Windows machines, but seems like can be done via: https://github.com/prometheus-community/windows_exporter.

2. **Permissions**: The smartctl binary requires elevated privileges to access device data.
   Grafana Alloy must run with one of the following:

   - Root permissions (not recommended for production)
   - `CAP_SYS_RAWIO` capability on Linux
   - Appropriate device permissions

> 3. **Kernel modules** (Linux only): Ensure appropriate kernel modules are loaded for your devices:

Reviewer: ```suggestion
3. On Linux only, make sure the appropriate kernel modules are loaded for your devices:
```
